src/main/kotlin/com/nikitanazarov/coroutinestacks/ui/CoroutineStacksPanel.kt

```diff
@@ -0,0 +1,168 @@
+                    emptyText.text = CoroutineStacksBundle.message("should.be.stopped.on.a.breakpoint")
+                }
+
+                override fun sessionCreated(session: DebuggerSession) {
+                    session.process.addDebugProcessListener(panelBuilderListener)
+                }
+
+                override fun sessionRemoved(session: DebuggerSession) {
+                    emptyText.text = CoroutineStacksBundle.message("no.java.debug.process.is.running")
+                    emptyText.component.isVisible = true
+                    panelContent.removeAll()
+                }
+            })
+    }
+
+    fun buildCoroutineGraph(suspendContext: SuspendContext) {
+        val suspendContextImpl = suspendContext as? SuspendContextImpl ?: run {
+            emptyText.text = CoroutineStacksBundle.message("coroutine.stacks.could.not.be.built")
+            return
+        }
+
+        val coroutineInfoCache: CoroutineInfoCache
+        try {
+            coroutineInfoCache = CoroutineDebugProbesProxy(suspendContextImpl).dumpCoroutines()
+        } catch (e: Exception) {
+            emptyText.text = CoroutineStacksBundle.message("nothing.to.show")
+            return
+        }
+
+        val coroutineInfoDataList = coroutineInfoCache.cache
+        val dispatcherToCoroutineDataList = mutableMapOf<String, MutableList<CoroutineInfoData>>()
+        for (info in coroutineInfoDataList) {
+            val dispatcher = info.descriptor.dispatcher ?: continue
+            dispatcherToCoroutineDataList.computeIfAbsent(dispatcher) { mutableListOf() }.add(info)
+        }
+
+        val dispatcherLabel = JLabel(CoroutineStacksBundle.message("select.dispatcher"))
+        val coroutineStacksPanelHeader = Box.createHorizontalBox()
+        val dispatchers = dispatcherToCoroutineDataList.keys.toTypedArray()
+        val dispatcherDropdownMenu = ComboBox(dispatchers)
+
+        dispatcherDropdownMenu.addActionListener {
+            val selectedDispatcher = dispatcherDropdownMenu.selectedItem as? String
+            val coroutineDataList = dispatcherToCoroutineDataList[selectedDispatcher]
+            if (!coroutineDataList.isNullOrEmpty()) {
+                updateCoroutineStackForest(coroutineDataList, suspendContextImpl)
+            }
+        }
+
+        val comboBoxSize = Dimension(comboBoxWidth, comboBoxHeight)
+        dispatcherDropdownMenu.apply {
+            preferredSize = comboBoxSize
+            maximumSize = comboBoxSize
+            minimumSize = comboBoxSize
+        }
+        coroutineStacksPanelHeader.add(dispatcherLabel)
+        coroutineStacksPanelHeader.add(dispatcherDropdownMenu)
+        val breakpointLocation = suspendContextImpl.location
+
+        dispatcherToCoroutineDataList.forEach { (dispatcher, coroutineDataList) ->
+            coroutineDataList.forEach { data ->
+                if (data.descriptor.state == State.RUNNING && data.stackTrace.first().location == breakpointLocation) {
```

Comment:
If `data.stackTrace` is empty we will get an exception here. It is better to use firstOrNull().